### ISSUE TITLE
Replace dim_map with stride_map

### ISSUE THREAD

**Issue by tardieu** [torch-spyre/torch-spyre#1044]
The current `SpyreTensorLayout` relies on a `dim_map` to map device dimensions to host dimensions.

The host `size` and `stride` of a tensor however are not guaranteed to remain constant post allocation as PyTorch may for instance squeeze and unsqueeze a tensor by altering its host `size` and `stride` after the tensor has been transferred to the device, hence given a `SpyreTensorLayout` and `dim_map`.

This behavior breaks the `dim_map` logic. The current `torch-spyre` code heuristically fixes the `dim_map` by making assumptions about the order of dimensions, which is fragile.

Moreover the current `dim_map` approach has a number of limitations:
- It cannot specify the order of dimensions within a tiled host dimension (#996).
- It does not support mapping one device dimension to multiple fused host dimensions.
- It does not support dimensions with stride 0, hence `torch.expand`.

We should explore replacing the `dim_map` with a `stride_map` to map device dimensions to host strides (ie offsets into the host tensor memory) instead of host dimensions. Offsets cannot be altered by any view operation since by definition a view does not alter the tensor storage.

Relying on offsets (from the tensor base address) is consistent with LLIR index functions, which similarly map iteration variables to host offsets. Using host offsets to reason about both host and device tensors also helps with code reuse when working with index functions.

A `stride_map` can support device dimensions that have no corresponding host dimensions by setting `stride_map[dim]=-1` as well as expanded dimensions with `stride_map[dim]=0`.

Strides in `stride_map` are not required to be host strides but simply host offsets making it possible to split or fuse dimensions as well as controlling the relative order of dimensions in tiled layouts.
